### PR TITLE
3950 ensure to close worker pools

### PR DESCRIPTION
**Pull request by akaspin**

### CODE REVIEW

File: group_processor.go [hunk 1]
```diff
@@ -166,7 +166,8 @@ func (gp *GroupProcessor) saveWorker(w *wp.Worker) {
 		case msg, ok := <-w.Channel():
 			if !ok {
 				gp.log.Warn("trying to read from closed worker channel")
-				continue
+				w.Done()
```
Comment (itszootime) [inline group_processor.go]: Suggestion: break here and call `w.Done()` before the function returns. From checking the logs, you can see that it never actually happens that it tries to read from a closed worker channel.

File: group_processor.go [hunk 2]
```diff
@@ -176,21 +177,23 @@ func (gp *GroupProcessor) saveWorker(w *wp.Worker) {
...
-	go func() {
-		gp.loadPool.Wait()
-		gp.wg.Done()
-	}()
+
 	go func() {
 		gp.exitErr = gp.Processor.Wait()
-		gp.loadPool.Close()
-		gp.savePool.Close()
+		poolsClosedCh := make(chan struct{})
```
Comment (itszootime) [inline group_processor.go]: This looks like a hack for an underlying bug somewhere in the worker pools and/or the group processor usage of them. Do you think it makes sense to wait for the worker pools? The consumer group has already exited, so those pools won't be able to commit offsets anyway.

Comment (akaspin) [reply group_processor.go]: Worker pools uses unified channel to return results. Wait a bit is OK. But in future it's good idea to refactor work pools implementation.

Comment (akaspin) [reply group_processor.go]: Hmm. You are right. Let's remove wait.

File: group_processor.go (hunk 1 above)
Comment (akaspin) [reply group_processor.go]: It's just safety change

File: group_processor.go (hunk 2 above)
Comment (akaspin) [reply group_processor.go]: Remove shutdown of worker pools will cause infinite goroutine and panic during shutdown. Unfortunately worker pool `Close()` method waits for workers. I don't want to dig all our services *right now* to ensure that nothing is depends on `pool.Close()` blocking behaviour (It's good for separate card). So. The safest behaviour is initiate `pool.Close()` and use timeout.

Comment (itszootime) [reply group_processor.go]: Do we even need to close the worker pools? The [`GroupProcessor` `Close()` method does not](https://github.com/remerge/go-group_processor/blob/master/group_processor.go#L204-L209).

Comment (akaspin) [reply group_processor.go]: `ggp.Close()` initiates `processor.Close()` which in turn [unblocks](https://github.com/remerge/go-group_processor/pull/20/files#diff-40195051b6de18d10aa99d25572cef9fR185) `processor.Wait()`. Unblocked `processor.Wait()` unleashes both `pool.Close()`.

I found that best practice to "close" any entity in Go is avoid to put shutdown routines to `Close()` method directly. Instead this `Close()` method should just pull the trigger that unleashes shutdown and when unblocks something like `Wait()`. The benefits are that `Close()` can be safely called many times (our case) and have unblocking behaviour. These are just architectural considerations. But they makes life more-more easier.

**Comment by itszootime** [comment]
@akaspin why did you merge this? It was not approved. My suggestion for simplification is replicate what we do in the `GroupProcessor` `Close()`, i.e. not care about shutting down the worker pools:

```
func (gp *GroupProcessor) Run() {
	gp.savePool.Run()
	gp.loadPool.Run()

	gp.wg.Add(1)
	go func() {
		gp.exitErr = gp.Processor.Wait()
		gp.wg.Done()
	}()
}

func (gp *GroupProcessor) Wait() error {
	gp.wg.Wait()
	return gp.exitErr
}
```

**Comment by akaspin** [comment]
@itszootime Oh! Sorry. Just miscommunication. Will revert. But current behaviour will remains unsafe! Look at [my comment](https://github.com/remerge/go-group_processor/pull/20#discussion_r344134014). Here simplification is bad practice.

**Pull request merged**